### ISSUE TITLE
Crash with webseed torrents

### ISSUE THREAD

**Issue by qu1ck** [transmission/transmission#3561]
I was running sanitizer build and got this crash with webseed torrents:

```
../libtransmission/webseed.cc:481:14: runtime error: member access within address 0x61f0002fb080 which does not point to an object of type '(anonymous namespace)::tr_webseed'
0x61f0002fb080: note: object has invalid vptr
 04 00 80 4d  02 00 00 1e 00 00 00 00  80 00 00 00 d0 61 00 00  80 43 3b 00 10 61 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              invalid vptr
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../libtransmission/webseed.cc:481:14 in
=================================================================
==1069597==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61f0002fba90 at pc 0x000000927dcb bp 0x7fed0ddfc250 sp 0x7fed0ddfc248
READ of size 8 at 0x61f0002fba90 thread T2
    #0 0x927dca in (anonymous namespace)::ConnectionLimiter::taskFinished(bool) /home/quick/src/transmission/build/../libtransmission/webseed.cc:98:9
    #1 0x927dca in (anonymous namespace)::onPartialDataFetched(tr_web::FetchResponse const&) /home/quick/src/transmission/build/../libtransmission/webseed.cc:481:33
    #2 0x6614df in onWorkAvailable(int, short, void*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:170:9
    #3 0x7fed131a29ee  (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x239ee)                                                                                                                                                   #4 0x7fed131a328e in event_base_loop (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x2428e)                                                                                                                                 #5 0x66058a in libeventThreadFunc(tr_event_handle*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:199:5                                                                                                 #6 0x7fed1288fecf  (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xceecf)                                                                                                                                                      #7 0x7fed127a7ea6 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8ea6)                                                                                                                                           #8 0x7fed12a8bdee in clone (/lib/x86_64-linux-gnu/libc.so.6+0xfddee)

Address 0x61f0002fba90 is a wild pointer.
SUMMARY: AddressSanitizer: heap-buffer-overflow /home/quick/src/transmission/build/../libtransmission/webseed.cc:98:9 in (anonymous namespace)::ConnectionLimiter::taskFinished(bool)                                      Shadow bytes around the buggy address:
  0x0c3e80057700: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057710: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057720: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057730: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057740: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c3e80057750: fa fa[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057760: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057770: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057780: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e80057790: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3e800577a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
Thread T2 created by T0 here:
    #0 0x4906fa in pthread_create (/home/quick/src/transmission/build/daemon/transmission-daemon+0x4906fa)                                                                                                                     #1 0x7fed12890144 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xcf144)                          #2 0x65fee2 in tr_eventInit(tr_session*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:223:19                                                                                                           #3 0x573cad in tr_sessionInit(char const*, bool, tr_variant*) /home/quick/src/transmission/build/../libtransmission/session.cc:632:5                                                                                       #4 0x4d977c in daemon_start(void*, bool) /home/quick/src/transmission/build/../daemon/daemon.cc:744:15                                                                                                                     #5 0x53184e in dtr_daemon(dtr_callbacks const*, void*, bool, int*, tr_error**) /home/quick/src/transmission/build/../daemon/daemon-posix.cc:223:18                                                                         #6 0x4d84f2 in main /home/quick/src/transmission/build/../daemon/daemon.cc:982:37                                                                                                                                          #7 0x7fed129b4d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

==1069597==ABORTING
```

Built from 15056045b

**Comment by qu1ck** [comment]
Another one:

```
../libtransmission/webseed.cc:481:14: runtime error: member access within address 0x61f000155480 which does not point to an object of type '(anonymous namespace)::tr_webseed'
0x61f000155480: note: object has invalid vptr
 07 00 00 77  02 00 80 25 00 00 00 00  80 00 00 00 d0 61 00 00  00 e6 0e 00 10 61 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              invalid vptr
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../libtransmission/webseed.cc:481:14 in
=================================================================
==1840269==ERROR: AddressSanitizer: heap-use-after-free on address 0x61f000155e90 at pc 0x000000927dcb bp 0x7f21fdafc250 sp 0x7f21fdafc248
READ of size 8 at 0x61f000155e90 thread T2
    #0 0x927dca in (anonymous namespace)::ConnectionLimiter::taskFinished(bool) /home/quick/src/transmission/build/../libtransmission/webseed.cc:98:9
    #1 0x927dca in (anonymous namespace)::onPartialDataFetched(tr_web::FetchResponse const&) /home/quick/src/transmission/build/../libtransmission/webseed.cc:481:33
    #2 0x6614df in onWorkAvailable(int, short, void*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:170:9
    #3 0x7f2202e749ee  (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x239ee)
    #4 0x7f2202e7528e in event_base_loop (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x2428e)
    #5 0x66058a in libeventThreadFunc(tr_event_handle*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:199:5
    #6 0x7f2202561ecf  (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xceecf)
    #7 0x7f2202479ea6 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8ea6)
    #8 0x7f220275ddee in clone (/lib/x86_64-linux-gnu/libc.so.6+0xfddee)

0x61f000155e90 is located 2576 bytes inside of 3176-byte region [0x61f000155480,0x61f0001560e8)
freed by thread T2 here:
    #0 0x4d61cd in operator delete(void*) (/home/quick/src/transmission/build/daemon/transmission-daemon+0x4d61cd)
    #1 0x7fb452 in tr_swarm::removePeer(tr_peer*) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:438:9
    #2 0x7c6b82 in disconnect_helpers::(anonymous namespace)::closePeer(tr_peer*) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:2378:18
    #3 0x7c5696 in disconnect_helpers::closeBadPeers(tr_swarm*, long) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:2435:9
    #4 0x7c8602 in tr_peerMgr::reconnectPulse() /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:2500:13
    #5 0x7ca6c0 in tr_peerMgr::bandwidthPulse() /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:2600:5
    #6 0x7e83d7 in tr_peerMgr::bandwidthPulseMarshall(int, short, void*) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:611:15
    #7 0x7f2202e749ee  (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x239ee)

previously allocated by thread T2 here:
    #0 0x4d596d in operator new(unsigned long) (/home/quick/src/transmission/build/daemon/transmission-daemon+0x4d596d)
    #1 0x8203ff in tr_peerMsgsNew(tr_torrent*, peer_atom*, tr_peerIo*, void (*)(tr_peer*, tr_peer_event const*, void*), void*) /home/quick/src/transmission/build/../libtransmission/peer-msgs.cc:869:12
    #2 0x7cffd2 in createBitTorrentPeer(tr_torrent*, tr_peerIo*, peer_atom*, unsigned long) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:1134:18
    #3 0x7b2e94 in on_handshake_done(tr_handshake_result const&) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:1243:13
    #4 0x971565 in fireDoneFunc(tr_handshake*, bool) /home/quick/src/transmission/build/../libtransmission/handshake.cc:1072:26
    #5 0x96dd0f in tr_handshakeDone(tr_handshake*, bool) /home/quick/src/transmission/build/../libtransmission/handshake.cc:1081:26
    #6 0x973fe4 in readPeerId(tr_handshake*, evbuffer*) /home/quick/src/transmission/build/../libtransmission/handshake.cc:705:12
    #7 0x96ecfd in canRead(tr_peerIo*, void*, unsigned long*) /home/quick/src/transmission/build/../libtransmission/handshake.cc:990:19
    #8 0x794a75 in canReadWrapper(tr_peerIo*) /home/quick/src/transmission/build/../libtransmission/peer-io.cc:136:29
    #9 0x79f3ba in tr_peerIoTryRead(tr_peerIo*, unsigned long) /home/quick/src/transmission/build/../libtransmission/peer-io.cc:1056:17
    #10 0x79e4f0 in tr_peerIoFlush(tr_peerIo*, tr_direction, unsigned long) /home/quick/src/transmission/build/../libtransmission/peer-io.cc:1147:45
    #11 0x739e29 in tr_bandwidth::phaseOne(std::vector<tr_peerIo*, std::allocator<tr_peerIo*> >&, tr_direction) /home/quick/src/transmission/build/../libtransmission/bandwidth.cc:181:32
    #12 0x73a865 in tr_bandwidth::allocate(tr_direction, unsigned int) /home/quick/src/transmission/build/../libtransmission/bandwidth.cc:233:5
    #13 0x7c9e90 in tr_peerMgr::bandwidthPulse() /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:2570:29
    #14 0x7e83d7 in tr_peerMgr::bandwidthPulseMarshall(int, short, void*) /home/quick/src/transmission/build/../libtransmission/peer-mgr.cc:611:15
    #15 0x7f2202e749ee  (/usr/lib/x86_64-linux-gnu/libevent-2.1.so.7+0x239ee)

Thread T2 created by T0 here:
    #0 0x4906fa in pthread_create (/home/quick/src/transmission/build/daemon/transmission-daemon+0x4906fa)
    #1 0x7f2202562144 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xcf144)
    #2 0x65fee2 in tr_eventInit(tr_session*) /home/quick/src/transmission/build/../libtransmission/trevent.cc:223:19
    #3 0x573cad in tr_sessionInit(char const*, bool, tr_variant*) /home/quick/src/transmission/build/../libtransmission/session.cc:632:5
    #4 0x4d977c in daemon_start(void*, bool) /home/quick/src/transmission/build/../daemon/daemon.cc:744:15
    #5 0x53184e in dtr_daemon(dtr_callbacks const*, void*, bool, int*, tr_error**) /home/quick/src/transmission/build/../daemon/daemon-posix.cc:223:18
    #6 0x4d84f2 in main /home/quick/src/transmission/build/../daemon/daemon.cc:982:37
    #7 0x7f2202686d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

SUMMARY: AddressSanitizer: heap-use-after-free /home/quick/src/transmission/build/../libtransmission/webseed.cc:98:9 in (anonymous namespace)::ConnectionLimiter::taskFinished(bool)
Shadow bytes around the buggy address:
  0x0c3e80022b80: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022b90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022ba0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022bb0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022bc0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c3e80022bd0: fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022be0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022bf0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022c00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3e80022c10: fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa
  0x0c3e80022c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1840269==ABORTING
```

**Comment by qu1ck** [comment]
This one happened right after I deleted webseed torrent that wasn't finished downloading, maybe that action is related.

**Comment by ckerr** [comment]
Is this still an issue after #4451, e.g. in 3cee8bae36f9 or newer?

**Comment by qu1ck** [comment]
I haven't run into this lately but I also haven't recompiled since 4.0 beta2 which is before the commit you linked.

**Comment by ckerr** [comment]
I _think_ this is fixed by the link above, so I'm going to close this issue for now. But @qu1ck please open a new issue if I'm wrong :smile_cat:
